### ISSUE TITLE
Clarify text surrounding decimal scale constraint

### ISSUE THREAD

**Issue by dlurton** [amzn/ion-schema#27]
The [section for the decimal scale constraint](https://amzn.github.io/ion-schema/docs/spec.html#scale) contains the following text:  

> The minimum scale must be greater than or equal to `0`.

This is inconsistent with the [data model for decimals](http://speleotrove.com/decimal/damodel.html) that is linked to by the [Ion specification](http://amzn.github.io/ion-docs/docs/decimal.html).  The data model document states:

> "When a limit to the exponent applies, it must result in a balanced range of positive or negative numbers"

**Comment by popematt** [comment]
More details:

The Ion Schema spec describes something like a SQL DECIMAL scale. (See, for example, [T-SQL](https://docs.microsoft.com/en-us/sql/t-sql/data-types/precision-scale-and-length-transact-sql?view=sql-server-ver16).)

This is fundamentally different than an Ion decimal because SQL Decimals are _fixed-point_ numbers, whereas Ion Decimals are arbitrary-precision numbers. Fixed-point numbers must have a scale, since it is a property of the data _type_, but for arbitrary precision values, there is not strictly a "scale", but there is an exponent which is a property of the _value_. However, the Ion specification does not seem to explicitly define any "scale"—only an "exponent" for Decimals.

The Kotlin reference implementation mostly parses/validates the ISL syntax according to the spec, but when validating an Ion value, the scale used is the `BigDecimal` scale. For example, the Ion value `1d1` has a `BigDecimal` scale of `-1`, which leads to some strange behavior in `ion-schema-kotlin`.

| type | value | result
|- |- |-
| `{ scale: -1 }` | `1d1` | invalid schema
| `{ scale: range::[-1, 0] } ` | `1d1` | invalid schema
| `{ scale: range::[0, max] } ` | `1d1` | invalid value
| `{ scale: range::[min, 0] } ` | `1d1` | valid
| `{ scale: range::[min, exclusive::0] } ` | `1d1` | valid

Observations:
* There is no way to create a schema that allows decimals with only some positive exponents (eg. allowing `10d1`, but not `1d2`)
* `0` and `min` behave differently, even though `0` is the minimum allowed value for the scale range. Any solution we come up with needs to resolve that inconsistency.
* `exclusive::0` is allowed even though it should probably not be a valid max value for the range _if_ we strictly follow the spec

**Comment by popematt** [comment]
**Proposal for Ion Schema 2.0**

In the next major version of Ion Schema, replace `scale` with `exponent`. The `exponent` constraint validates the exponent of the Ion `decimal` data model, and can be constrained with a single number or a range.

**Proposal for Ion Schema 1.x**

Change ISL 1.0 `scale` to allow negative numbers.

**Comment by popematt** [comment]
In response to an offline conversation about SQL `DECIMAL` and how it relates to Ion Schema, here is some documentation on how the `exponent` constraint relates to a SQL `DECIMAL` understanding of _scale_. (All of these examples hold true for the current `scale` constraint, but the constraint argument must be inverted/negated. E.g. `exponent: range::[-2, max]` => `scale: range::[min, 2]`.)

## Modeling a SQL Decimal in Ion Schema 2.0

Ion `decimal` and SQL `DECIMAL` have some fundamental differences that make it complex to compare values between the two types.

For SQL `DECIMAL`, the precision of the number `12.34` is not determined by the actual number of digits, but (speaking informally) by the data type it is assigned. `12.34` can be assigned to `DECIMAL(38,2)` just as easily as `DECIMAL(4,2)`. That is to say, precision and scale are a property of the `DECIMAL` data _type_ rather than being inherent in the value.

On the other hand, the Ion decimal data type has no particular precision or scale as it is an _arbitrary precision_ data type—the precision of an Ion decimal value is inherent in the value itself. 

As a result of these differences, when converting a value from a SQL `DECIMAL` to an Ion `decimal`, it is _possible_ to preserve the scale of the `DECIMAL` (as it becomes the exponent of the Ion decimal, multiplied by -1), but the precision is lost. When converting from an Ion `decimal` to a `DECIMAL(p,s)`, the precision and scale of the Ion value are always lost because the value is coerced into having the precision and scale of `DECIMAL(p,s)`.

That being said, here are some questions that one might be trying to answer by modeling SQL `DECIMAL` and how to approach modeling them in Ion Schema.

**Does `x` have the exact precision and the exact scale of `DECIMAL(p,s)`?**

This is probably the least useful way to model a SQL `DECIMAL`, and is included here only to help explain why exact precision is not useful when modeling `DECIMAL` in Ion Schema.

Using `DECIMAL(5,2)` as our example, we can model an _exact_ precision and scale like this:
```
type::{
  precision: 5,
  exponent: -2,
}
```
This will only allow `decimal`s with 5 digits, and two after the decimal point. However, this will not accept valid `DECIMAL(5,2)` values such as `1.00`.

**Does `x` have a compatible precision and the exact scale of `DECIMAL(p,s)`?**

This way of modeling a SQL `DECIMAL` is useful for scenarios where the value was a SQL `DECIMAL(p,s)` before being converted to Ion.

This can be separated into two parts:
1. Does `x` have exactly `s` digits after the decimal point?
2. Does `x` have less than or equal to `p` digits? (Or is `-10^(p-s) < x < 10^(p-s)`?)


Using `DECIMAL(5,2)` as our example, we can model a _compatible_ precision and _exact_ scale like this:
```
type::{
  precision: range::[min, 5],
  exponent: -2,
}
```
This will accept values such as `100.00`, `1.50`, and `0.01`. It will reject values such as `100`, `1.5`, `0.010`, and `1000.0`.

Instead of using the `precision` constraint, we could also use `valid_values: range::[-999.99, 999.99]` to achieve the same result.

**Does `x` fit in a `DECIMAL(p,s)` without having to round or truncate any digits after the decimal point?**

This is probably the most useful way to model a SQL `DECIMAL` in Ion Schema. It can be used for validating that data _could_ have been a `DECIMAL(p,s)` before it was converted to Ion, and that the value could be converted to a `DECIMAL(p,s)` without any rounding or truncating.

This can be separated into two parts:
1. Does `x` have less than or equal to `s` digits after the decimal point?
2. Is `-10^(p-s) < x < 10^(p-s)`?

Using `DECIMAL(5,2)` as our example again, we can model a _compatible_ precision and scale like this:
```
type::{
  exponent: [-2, max],
  // Scale is not fixed, so we cannot use the `precision` constraint. Must use `valid_values` instead.
  valid_values: range::[ -999.99, 999.99 ],
}
```
This will accept values such as `1.5`, `1.50`, `5d2`, and `500`. It will reject values such as `1200`, `1.2d5`, `1.501`, and `1.500`.

**Comment by popematt** [comment]
Resolved by #76 in Ion Schema 2.0 and by adding documentation in #74.
